File: day05/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	ids := []int{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		n := calcID([]byte(scanner.Text()))
		ids = append(ids, n)
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	sort.Ints(ids)

	if len(ids) == 0 {
		log.Fatal("No ids")
	}

	fmt.Println("Part 1:", ids[len(ids)-1])

	prev := ids[0] - 1
	for _, i := range ids {
		if i-prev > 1 {
			fmt.Println("Part 2:", i-1)
			break
		}
		prev = i
	}
}

func calcID(b []byte) int {
	n := 0
	for _, i := range b {
		n *= 2
		if i == 'B' || i == 'R' {
			n++
		}
	}
	return n
}

File: day05/src/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let v = {
        let mut v = Vec::new();
        for line in reader.lines() {
            let line = line?;
            let chars = line.chars().collect::<Vec<_>>();
            v.push(calc_id(&chars));
        }
        v.sort();
        v
    };

    println!("Part 1: {}", v.last().ok_or("No ids")?);

    let mut prev = v.first().ok_or("No ids")?;
    for i in &v {
        if i - prev > 1 {
            println!("Part 2: {}", i - 1)
        }
        prev = i
    }

    Ok(())
}

fn calc_id(b: &Vec<char>) -> usize {
    let mut n = 0;
    for &i in b {
        n *= 2;
        if i == 'B' || i == 'R' {
            n += 1;
        }
    }
    n
}
